src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.Data.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.NetCore.Analyzers.InteropServices
+{
+    public sealed partial class PlatformCompatabilityAnalyzer
+    {
+        /// <summary>
+        /// Class used for keeping platform information of an API, all are optional properties
+        /// 
+        /// We need to keep only 2 values for [SupportedOSPlatform] attribute, first one will be the lowest version found, mostly for assembly level attribute which denotes when the API first introduced,
+        /// second one would keep new APIs added later and requries higher platform version (if there is multiple version found in the API parents chain we will keep only highest version)
```

Comment:
> (if there is multiple version found in the API parents chain we will keep only highest version)

Today, I discussed this with @jeffhandley. My belief was that chaining from child to parent wouldn't be sound because we have two models: the allow list and the deny list. But while I was typing that comment into this text box,, I think I convinced myself that we need to merge attributes. For example:

```C#
[SupportedOSPlatform("ios10.0")]
[SupportedOSPlatform("tvos4.0")]
class Some
{
    public void Api1() {}

    [UnsupportedOSPlatform("tvos")]
    public void Api2() {}
}
```

`Api1()` is supported on iOS and tvOS but `Api2()` is only supported on iOS. I would argue that it's nonsensical if a parent could declare that it works on fewer platforms than some of its children. But if we have the rule that I suggested, that is you only consider the attributes of the first API, than we would conclude that `Api2()` works on all platforms (including Windows, MacOS, etc) except for tvOS. I'd argue that this doesn't make any sense here. Instead, we should conclude that `Api2()` is only supported on iOS.

Stated differently: the rules should be such a child restrict the set of platforms, but not expand.

When evaluating call sites, we should perform the following walk:

* If an API uses the allow-list model, then we don't need to walk any further. An allow list means anything not stated, isn't included.
* If an API uses the deny-list model, then we must keep walking. Further deny-list entries can subtract more platforms. Once we hit an allow-list API, we can stop and compute the effective set of APIs to be everything in the allow-list minus all the accumulated deny-list entries.

For the other analyze that checks declarations of well-formedness we should:

* Validate platform name and version syntax (#4015)
* Validate the platform name to be a known one (#4014)
* Check that a child API only restricts the platform set, and never expands (#4022, just filed)

/cc @bartonjs

Reply:
Good rationale.  If we took that approach, then we could close #4020. @buyaa-n is this possible to pull off before Monday morning?

Reply:
@terrajobst thanks for the clear definition

> @buyaa-n is this possible to pull off before Monday morning

That is what i am aiming for, as @manish has resolved some pending logic now this will be prioritized first for me

Reply:
For the following scenario, the effective set is still undefined with the logic above:

```csharp
    [SupportedOSPlatform("windows")]
    class TargetSupportedOnWindows
    {
        [UnsupportedOSPlatform("windows")]
        public static void FunctionUnsupportedOnWindows() { }
    }
```

@buyaa-n and I agreed that `[UnsupportedOSPlatform]` should override the parent's declaration of `[SupportedOSPlatform]`

Given the inverse scenario, since the API applies an allow-list on the API, the effective result is `[SupportedOSPlatform]` wins because we don't even perform the walk.

```csharp
    [UnsupportedOSPlatform("windows")]
    class TargetUnsupportedOnWindows
    {
        [SupportedOSPlatform("windows")]
        public static void FunctionSupportedOnWindows() { }
    }
```